manager/api/transport_test.go

```diff
@@ -0,0 +1,535 @@
+		validCredentials, _ := json.Marshal(validUser)
+		invalidEmailCredentials, _ := json.Marshal(invalidUser)
+		invalidCredentials, _ := json.Marshal(invalidPassUser)
+		nonexistentCredentials, _ := json.Marshal(nonexistentUser)
+		svc.Register(validUser)
+
+		cases := []struct {
+			desc   string
+			req    string
+			status int
+			res    string
+		}{
+			{"login with valid credentials", string(validCredentials), http.StatusCreated, string(token)},
+			{"try to login with invalid credentials", string(invalidCredentials), http.StatusForbidden, ""},
+			{"try to login with invalid email address", string(invalidEmailCredentials), http.StatusBadRequest, ""},
+			{"try to login non-existent user", string(nonexistentCredentials), http.StatusForbidden, ""},
+			{"try to login with invalid request format", "{", http.StatusBadRequest, ""},
+		}
+
+		for _, tc := range cases {
+			res, err := sendRequest(client, http.MethodPost, fmt.Sprintf("%s/tokens", url), contentType, "", strings.NewReader(tc.req))
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+			body, err := ioutil.ReadAll(res.Body)
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+			token := strings.Trim(string(body), "\n")
+
+			assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+			assert.Equal(t, tc.res, token, fmt.Sprintf("%s: expected body %s got %s", tc.desc, tc.res, token))
+		}
+	})
+}
+
+func TestAddClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		validClient, _ := json.Marshal(validClient)
+		invalidClient, _ := json.Marshal(invalidClient)
+		svc.Register(validUser)
+
+		cases := []struct {
+			desc   string
+			req    string
+			auth   string
+			status int
+		}{
+			{"add valid client", string(validClient), validUser.Email, http.StatusCreated},
+			{"try to add client with invalid data", string(invalidClient), validUser.Email, http.StatusBadRequest},
+			{"try to add client with invalid auth token", string(validClient), "invalid_token", http.StatusForbidden},
+			{"try to add client with invalid request format", "}", validUser.Email, http.StatusBadRequest},
+		}
+
+		for _, tc := range cases {
+			res, err := sendRequest(client, http.MethodPost, fmt.Sprintf("%s/clients", url), contentType, tc.auth, strings.NewReader(tc.req))
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+			assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+		}
+	})
+}
+
+func TestUpdateClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		validUpdate, _ := json.Marshal(manager.Client{
+			Type:    "device",
+			Name:    "test_device",
+			Payload: "device_payload",
+		})
+		invalidUpdate, _ := json.Marshal(manager.Client{
```

Comment:
Create invalid based on the existing valid. This will make the "error" more explicit.